# Introduction to Software Testing
>原文：https://www.fuzzingbook.org/html/Intro_Testing.html
在进入书的中心部分之前,让我们先介绍软件测试的基本概念。为什么软件测试是必要的?怎样测试软件?怎样判断一个测试是否成功?怎样知道测试是否充分?在本章中,我们先回顾最重要的概念,同时也熟悉一下Python和交互笔记本。

## Simple Testing

从一个简单的例子开始，假设要实现一个平方根函数，使用 [Newton–Raphson method](https://en.wikipedia.org/wiki/Newton's_method)计算方法，，用Pyhton实现`my_sqrt()`函数计算平方根：

```c
def my_sqrt(x):
    approx = None
    guess = x / 2
    while approx != guess:
        approx = guess
        guess = (approx + x / approx) / 2
    return approx
```

确认这个函数是否真的实现了上述功能

## Understanding Python Programs

如果没学过Python，知道下面几点：

1. Python通过缩进来组织程序结构,所以函数体和`while`循环体是通过缩进来定义的;
2. Python是动态类型的,这意味着像`x`、`approx`或`guess`这样的变量类型是在运行时确定的。

## Running a Function

为了测试`my_sqrt()`是否正常工作，我们可以使用一些值对其进行测试。例如`x = 4`

```c
my_sqrt(4)
2.0
```

测试`x = 2.0`：

```c
my_sqrt （2 ）
1.414213562373095
```



## Debugging a Function

了解`my_sqrt()`内部运行方式，一种简单的方法是在关键位置插入`print()`语句。例如，您可以记录值`approx`，以查看每次循环迭代如何逐步接近实际值：

```c
def my_sqrt_with_log(x):
    approx = None
    guess = x / 2
    while approx != guess:
        print("approx =", approx)  # <-- New
        approx = guess
        guess = (approx + x / approx) / 2
    return approx
```
测试`my_sqrt_with_log(9)`
```c
approx = None
approx = 4.5
approx = 3.25
approx = 3.0096153846153846
approx = 3.000015360039322
approx = 3.0000000000393214
3.0
```



## Checking a Function

上面的`my_sqrt(2)`计算的答案真的正确吗？我们可以通过以下方式来验证：

```c
my_sqrt(2) * my_sqrt(2)
1.9999999999999996
```

应该是浮点数精度的问题导致



## Automating Test Execution

我们已经手动测试了上述程序，手动运行它并手动检查其结果。这是一种非常灵活的测试方法，但是从长远来看，它效率很低：

1. 手动测试只能检查非常有限的部分功能及其结果
2. 对程序进行任何更改后，必须重复测试过程

一种简单的方法是让计算机首先进行计算，然后让计算机检查结果，下面这段代码测试了`sqrt(4)=2`的正确性

```c
result = my_sqrt(4)
expected_result = 2.0
if result == expected_result:
    print("Test passed")
else:
    print("Test failed")
```

```c
Test passed
```

这个测试的好处是我们可以一次又一次地运行它，从而确保至少正确计算了4的平方根。但是，仍然存在许多问题：

1. 单个测试需要五行代码
2. 它不在乎舍入错误
3. 它仅检查单个输入（和单个结果）

一一解决上面的问题，首先让测试更加紧凑，几乎所有语言都提供了`断言`---自动检查条件是否成立，如果条件不成立则立即停止执行，

在Python中，我们使用`assert`语句，如果条件为true，则什么也不会发生。（如果一切正常，则不应该出问题。）但是，如果条件评估为false，`assert`则会引发异常，表明测试刚刚失败。

```c
assert my_sqrt(4) == 2
```

当您执行此行代码时，什么都不会发生：我们证明（或断言）了上面的计算正确

但是，浮点计算可能会导致舍入误差，因此，我们不能简单地比较两个浮点数是否相等。因此，我们将确保它们之间的绝对差保持在某个阈值以下，通常表示为或`epsilon`。这是我们可以做到的：

```c
EPSILON = 1e-8
assert abs(my_sqrt(4) - 2) < EPSILON
```

还可以介绍一个特殊函数`abs()`求绝对值，进行更多的测试

> 下面这个函数用更巧妙的方式进行断言：判断`计算后的值`和`正确的值`的差是否小于设定的阈值来判断计算是否正确

```c
def assertEquals(x, y, epsilon=1e-8):
    assert abs(x - y) < epsilon
```

```c
assertEquals(my_sqrt(4), 2)
assertEquals(my_sqrt(9), 3)
assertEquals(my_sqrt(100), 10)
```

## Generating Tests

还可以使用下面的思路进行测试：
$$
\sqrt{x} * \sqrt{x}
$$

```c
assertEquals(my_sqrt(2) * my_sqrt(2), 2)
assertEquals(my_sqrt(3) * my_sqrt(3), 3)
assertEquals(my_sqrt(42.11) * my_sqrt(42.11), 42.11)
```

可以通过下面的方式测试成千上万的值

```c
for n in range(1, 1000):
	assertEquals(my_sqrt(n) * my_sqrt(n), n)
```

使用自己的Timer模块来计算花费的时间，

```c
import bookutils
from Timer import Timer
with Timer() as t:
    for n in range(1, 10000):
        assertEquals(my_sqrt(n) * my_sqrt(n), n)
print(t.elapsed_time())
//0.01560891700501088
```

10,000个值大约需要百分之一秒，因此单次执行`my_sqrt()`需要1/1000000秒或大约1微秒。

让我们重复随机选择10,000个值。Python的`random.random()`函数返回0.0到1.0之间的随机值：

```c
import random
with Timer() as t:
    for i in range(10000):
        x = 1 + random.random() * 1000000
        assertEquals(my_sqrt(x) * my_sqrt(x), x)
print(t.elapsed_time())
//0.01828445799765177
```

一秒钟之内，测试了10,000个随机值，并且每次实际上都正确计算了平方根。只要对进行任何更改`my_sqrt()`，我们就可以重复进行此测试，每次都可以增强我们`my_sqrt()`没有隐患的信心。



## Run-Time Verification

除了人为编写和运行测试`my_sqrt()`，我们还可以将检查权限集成到实现中。 这样，每次运行都会调用`-3`进行自动检查。

这样的自动运行时检查非常容易实现：

```c
def my_sqrt_checked(x):
    root = my_sqrt(x)
    assertEquals(root * root, x)
    return root
```

这样，无论什么时候用`my_sqrt_checked()`进行计算都可以确保结果是正确的。因为在计算后进行了检查。

自动运行时检查 基于 以下2点假设:

1. 要能够构建这样的运行时检查。应该总是能够得到具体的值来检查,但是用抽象的方式来表达所需的性质可能非常复杂。在实践中,你需要决定哪些性质最关键,并为它们设计适当的检查。另外,运行时检查可能不仅依赖于局部性质,还依赖程序状态的多个属性,所有这些都需要识别出来。
2. 要能够承受这样的运行时检查。对于`my_sqrt()`来说,检查的代价不是很高;但是如果我们需要在一个简单的操作之后检查一个大的数据结构,检查的成本很快就会变得过高。在实践中,运行时检查通常会在生产环境中被禁用,以效率换取可靠性。另一方面,一个全面的运行时检查套件是发现错误和快速调试它们的好方法;你需要决定在生产环境下还需要多少这类功能。

运行时检查的一个重要局限是,它们只能在有结果可以检查时确保正确性——也就是说,它们不能保证结果总是存在。这与符号验证技术和程序证明形成明显对比,后者即使付出更大的(通常是手动的)努力,也可以保证结果的存在。

## System Input vs Function Input

在这一部分，我们会把`my_sqrt()`提供给其他程序员，然后他们可以将其嵌入他们的代码中。在某些时候，它必须处理来自第三方用户的输入，即不受程序员的控制。

让我们通过假设一个程序的输入为第三方字符串来模拟此系统输入：`sqrt_program()`

```c
def sqrt_program(arg: str) -> None:
    x = int(arg)
    print('The root of', x, 'is', my_sqrt(x))
```

调用`sqrt_program()`

```c
sqrt_program("4")
The root of 4 is 2.0
```

上面的代码存在问题，加入传入的参数是负数呢？试一下，`sqrt_program(-1)`，代码会进入无线循环。可以使用`with ExpectTimeOut(1)`让代码在一秒钟后中断执行、

```c
from ExpectError import ExpectTimeout

with ExpectTimeout(1):
sqrt_program("-1")
```

```c
Traceback (most recent call last):
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_935/1288144681.py", line 2, in <cell line: 1>
    sqrt_program("-1")
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_935/449782637.py", line 3, in sqrt_program
    print('The root of', x, 'is', my_sqrt(x))
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_935/2661069967.py", line 5, in my_sqrt
    while approx != guess:
  File "/Users/zeller/Projects/fuzzingbook/notebooks/Timeout.ipynb", line 43, in timeout_handler
    raise TimeoutError()
TimeoutError (expected)
```

上面是错误消息，

它列出了错误发生时处于活动状态的函数的调用堆栈。最底部的行是最后执行的行；上面的几行代表函数调用

们不希望我们的代码以异常终止。因此，在接受外部输入时，我们必须确保已对其进行正确验证。例如我们可以写：

```c
def sqrt_program(arg):
    x = int(arg)
    if x < 0:
        print("非法输入")
    else:
        print('The root of', x, 'is', my_sqrt(x))
```

新的问题又出现了，如果`sqrt_program()`不使用数字调用怎么办？当我们试图转换一个非数值的字符串时,这也会导致运行时错误:

```c
from ExpectError import ExpectError

with ExpectError():
sqrt_program("xyzzy")
```

报错信息如下：

```c
Traceback (most recent call last):
File "<ipython-input-29-8c5aae65a938>", line 2, in <module>
sqrt_program("xyzzy")
File "<ipython-input-26-ea86281b33cf>", line 2, in sqrt_program
x = int(arg)
ValueError: invalid literal for int() with base 10: 'xyzzy' (expected)
```

改代码，加入新的检查

```PY
def sqrt_program(arg):
    try:
        x = float(arg)
    except ValueError:
        print("Illegal Input")  #如果arg不是一个可转换的数字字符串,会触发ValueError异常。
    else:
        if x < 0:
            print("Illegal Number")
        else:
            print('The root of', x, 'is', my_sqrt(x))
```

在系统级别,程序必须保证能够正常地处理任何类型的输入，永远不会进入不受控制的状态。这无疑为程序员带来了负担,他们必须努力使程序健壮到能处理各种情况。然而,在生成软件测试时,这种负担却变成了好处:如果一个程序可以处理任何输入(可能伴随明确的错误信息),我们也可以给它任何输入。



## The Limits of Testing

尽管已经在不断改进测试，但请记住我们只是始终在检查功能是否有某种错误的输入。因此，可能总是存在未经测试的输入仍会导致失败。

> 永远做不到穷举所有的错误输入。

例如：

```c
with ExpectError():
root = my_sqrt(0)
```

报错如下：

```c
Traceback (most recent call last):
File "<ipython-input-34-24ede1f53910>", line 2, in <module>
root = my_sqrt(0)
File "<ipython-input-1-47185ad159a1>", line 7, in my_sqrt
guess = (approx + x / approx) / 2
ZeroDivisionError: float division by zero (expected)
```

又加一条检查：

```c
def my_sqrt_fixed(x):
    assert 0 <= x
    if x == 0:
        return 0
    return my_sqrt(x)
```

---

尽管广泛的测试可以使我们对程序的正确性有很高的信心，但它不能保证所有将来的执行都是正确的。即使是运行时验证,它检查每个结果,也只能保证如果它产生了一个结果,这个结果就是正确的;但是不能保证未来的执行不会导致检查失败。

## Lessons Learned

- 测试的目的是执行程序以发现错误。
- 测试执行、测试生成和检查测试结果可以自动化。
- 测试是不完整的;它不能100%保证代码没有错误。

## Next Steps

接下来，可以阅读

[use fuzzing to test programs with random inputs]()



## Background(拓展阅读)

There is a large number of works on software testing and analysis.

- An all-new modern, comprehensive, and online textbook on testing is ["Effective Software Testing: A Developer's Guide"](https://www.effective-software-testing.com/) [[Maurício Aniche, 2022](https://www.effective-software-testing.com/)]. Much recommended!
- For this book, we are also happy to recommend "Software Testing and Analysis" [[Pezzè *et al*, 2008](http://ix.cs.uoregon.edu/~michal/book/)] as an introduction to the field; its strong technical focus very well fits our methodology.
- Other important must-reads with a comprehensive approach to software testing, including psychology and organization, include "The Art of Software Testing" [[Myers *et al*, 2004](https://dl.acm.org/citation.cfm?id=983238)] as well as "Software Testing Techniques" [[Beizer *et al*, 1990](https://dl.acm.org/citation.cfm?id=79060)].

# Exercises

## Exercise 1: Get Acquainted with Notebooks and Pytho

第一个练习是熟悉notebooks 和 Python

## Exercise 2: Testing Shellsort

请看下面`shellsort`函数的实现，它获取一个元素列表并对其进行排序。

```c
def shellsort(elems):
    sorted_elems = elems.copy()
    gaps = [701, 301, 132, 57, 23, 10, 4, 1]
    for gap in gaps:
        for i in range(gap, len(sorted_elems)):
            temp = sorted_elems[i]
            j = i
            while j >= gap and sorted_elems[j - gap] > temp:
                sorted_elems[j] = sorted_elems[j - gap]
                j -= gap
            sorted_elems[j] = temp

    return sorted_elems
```

第一次测试：

```c
shellsort([3, 2, 1])
[1, 2, 3]
```

好像函数实现了上述功能。上面的实现使用了一个列表`elems`作为参数(它被复制到`sorted_elems`)，以及固定列表`gaps`，列表在Python中的工作方式类似其他语言中的数组:

```c
a = [5, 6, 99, 7]
print("First element:", a[0], "length:", len(a))
First element: 5 length: 4
```
`range()`函数返回一个可迭代的元素列表。它通常与for循环一起使用,如上面的实现所示。

```c
for x in range(1, 5):
    print(x)
1
2
3
4
```

---

**Part 1: Manual Test Cases**

现在,您的任务是用各种输入来彻底测试`shellsort()`。

首先,设置多个手动编写的测试用例的`assert`语句。选择测试用例以覆盖极端情况。使用`==`来比较两个列表

给一些测试用例：

```c
# 标准列表
assert shellsort([3, 2, 1]) == [1, 2, 3]
assert shellsort([1, 2, 3, 4]) == [1, 2, 3, 4]
assert shellsort([6, 5]) == [5, 6]
```


```c
# 重复元素的列表
assert shellsort([2, 2, 1]) == [1, 2, 2]
```

```c
# 空列表
assert shellsort([]) == []
```

我这里是通过print来确认哪组数据出现问题，结果上述测试用例都通过：

![image](https://github.com/kksaert/fuzzbook/assets/83335903/278e78e8-8ae4-41e7-b490-32b0ce981b31)


**Part 2: Random Inputs**

其次,创建随机列表作为`shellsort()`的参数。利用以下代码来检查结果(a)是否已排序,以及(b)是否是原始列表。

```c
def is_sorted(elems):
    return all(elems[i] <= elems[i + 1] for i in range(len(elecms) - 1))
```

```c
is_sorted([3, 5, 9])  #返回True
```

```c
def is_permutation(a, b):
    return len(a) == len(b) and all(a.count(elem) == b.count(elem) for elem in a)
```

> 用了all函数和生成器表达式：遍历a中的每个元素elem,看它在a和b中的出现次数是否相同。所有元素都满足这个条件,才返回True。

```c
is_permutation([3, 2, 1], [1, 3, 2]) # 返回True
```

---

创建一个随机列表生成器,使用`[]`作为空列表,使用`elems.append(x)`来将元素`x`附加到列表`elems`。使用上述帮助函数来评估结果。生成并测试1000个列表。

下面的函数做到了两个随机，列表随机出现长度(1到10)，列表内容随机。

```c
def random_list():
    length = random.randint(1, 10)
    elems = []
    for i in range(length):
        elems.append(random.randint(0,1000))
    return elems 
```

进行测试

![image](https://github.com/kksaert/fuzzbook/assets/83335903/3c19096a-b04a-4a0a-8c90-c863be0190f4)



对`shellsort()`进行1000次随机测试

```c
for i in range(1000):
    elems = random_list()
    sorted_elems = shellsort(elems)
    assert is_sorted(sorted_elems) and is_permutation(sorted_elems, elems)
```

测试结束，这里没有精确计算运行时间

![image](https://github.com/kksaert/fuzzbook/assets/83335903/16caadbf-9a15-4f2f-a17e-128d07ea606c)


## Exercise 3: Quadratic Solver

给了一个数学公式，用代码实现如下：

![Uploading image.png…]()


```c
def quadratic_solver(a, b, c):
    q = b * b - 4 * a * c
    solution_1 = (-b + my_sqrt_fixed(q)) / (2 * a)
    solution_2 = (-b - my_sqrt_fixed(q)) / (2 * a)
    return (solution_1, solution_2)
```

```c
quadratic_solver(3, 4, 1)
>(-0.3333333333333333, -1.0)
```

上面的代码实现存在漏洞

1. 除以0
2. 违反`my_sqrt_fixed`函数

如何避免上述问题？

**Part 1: Find bug-triggering inputs**

针对上述两种情况：找出能引发问题的a,b,c的值：

第一种情况，报错如下：

```c
with ExpectError():
    print(quadratic_solver(0, 0, 1))
```

```c
Traceback (most recent call last):
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_935/273293533.py", line 2, in <cell line: 1>
    print(quadratic_solver(0, 0, 1))
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_935/597421383.py", line 3, in quadratic_solver
    solution_1 = (-b + my_sqrt_fixed(q)) / (2 * a)
ZeroDivisionError: division by zero (expected)
```

第二种情况，报错如下：

```c
with ExpectError():
    print(quadratic_solver(3, 2, 1))
```

```c
Traceback (most recent call last):
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_935/284450257.py", line 2, in <cell line: 1>
    print(quadratic_solver(3, 2, 1))
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_935/597421383.py", line 3, in quadratic_solver
    solution_1 = (-b + my_sqrt_fixed(q)) / (2 * a)
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_935/3001478627.py", line 2, in my_sqrt_fixed
    assert 0 <= x
AssertionError (expected)
```

---

**Part 2: Fix the problem**

适当扩展代码，以便处理这些情况。对于不存在的值，返回 None。

加两个判断即可(第一个代码是自己写的，第二个是作者给的)：

```c
def quadratic_solver(a, b, c):
    q = b * b - 4 * a * c
    if a > 0 and q >=0:
        solution_1 = (-b + my_sqrt_fixed(q)) / (2 * a)
        solution_2 = (-b - my_sqrt_fixed(q)) / (2 * a)
        return (solution_1, solution_2)
```

```c
def quadratic_solver_fixed(a, b, c):
    if a == 0:
        if b == 0:
            if c == 0:
                # Actually, any value of x
                return (0, None)
            else:
                # No value of x can satisfy c = 0
                return (None, None)
        else:
            return (-c / b, None)

    q = b * b - 4 * a * c
    if q < 0:
        return (None, None)

    if q == 0:
        solution = -b / 2 * a
        return (solution, None)

    solution_1 = (-b + my_sqrt_fixed(q)) / (2 * a)
    solution_2 = (-b - my_sqrt_fixed(q)) / (2 * a)
    return (solution_1, solution_2)
```

作者似乎是对a,b，c三个值都进行了判断，更保险吧。

---

**Part 3: Odds and Ends**

随机输入发现最开始的两个问题的机会有多大?假设每秒可以进行10亿次测试,要平均等待多久一个错误才会被触发?

如果a，b,c的取值范围可以达到2的32次方，那么仅第一个条件,`a`和`b`同时为零,其概率为`p=1/(2^32*2^32)=1/2^64`,也就是184万亿分之一:

如果我们每秒能进行十亿次测试，我们需要等待多少年？

```c
tests_per_second = 1000000000
seconds_per_year = 60 * 60 * 24 * 365.25
tests_per_year = tests_per_second * seconds_per_year
combinations / tests_per_year
```

```c
584.5420460906264
```

我们看到，平均需要等待 584 年。显然，纯粹的随机选择作为唯一的测试策略是不够的。



## Exercise 4: To Infinity and Beyond

`my_sqrt_fixed(x)` 对所有有限数x都有效时，如果将设置x为无穷大会怎样？试试这个！

```c
from ExpectError import ExpectTimeout

infinity = float('inf')  # 这种方法得到一个无限大的数
with ExpectTimeout(1):
    y = my_sqrt_fixed(infinity)
```

报错如下：可以看到是一个超时错误

```c
Traceback (most recent call last):
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_935/3033877499.py", line 4, in <cell line: 3>
    y = my_sqrt_fixed(infinity)
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_935/3001478627.py", line 5, in my_sqrt_fixed
    return my_sqrt(x)
  File "/var/folders/n2/xd9445p97rb3xh7m1dfx8_4h0006ts/T/ipykernel_935/2661069967.py", line 5, in my_sqrt
    while approx != guess:
  File "/Users/zeller/Projects/fuzzingbook/notebooks/Timeout.ipynb", line 43, in timeout_handler
    raise TimeoutError()
TimeoutError (expected)
```

> 如果用数字测试程序，一定要确保包含极端值。
